util/limiter/limiter.go

```diff
@@ -0,0 +1,126 @@
+//
+// Each key gets its own separate token bucket to pull from, enabling
+// enforcement on things like "requests per IP address". To avoid
+// unbounded memory growth, Limiter actually only tracks limits
+// precisely for the N most recently seen keys, and assumes that
+// untracked keys are well-behaved. This trades off absolute precision
+// for bounded memory use, while still enforcing well for outlier
+// keys.
+//
+// As such, Limiter should only be used in situations where "rough"
+// enforcement of outliers only is sufficient, such as throttling
+// egregious outlier keys (e.g. something sending 100 queries per
+// second, where everyone else is sending at most 5).
+//
+// Each key's token bucket behaves like a regular token bucket, with
+// the added feature that a bucket's token count can optionally go
+// negative. This implements a form of "cooldown" for keys that exceed
+// the rate limit: once a key starts getting denied, it must stop
+// requesting tokens long enough for the bucket to return to a
+// positive balance. If the key keeps hammering the limiter in excess
+// of the rate limit, the token count will remain negative, and the
+// key will not be allowed to proceed at all. This is in contrast to
+// the classic token bucket, where a key trying to use more than the
+// rate limit will get capped at the limit, but can still occasionally
+// consume a token as one becomes available.
+type Limiter[K comparable] struct {
+	// Size is the number of keys to track. Only the Size most
+	// recently seen keys have their limits enforced precisely, older
+	// keys are assumed to not be querying frequently enough to bother
+	// tracking.
+	Size int
+
+	// Burst is the number of tokens available for a key to consume
+	// before time-based rate limiting kicks in. An unused limiter
+	// regains available tokens over time, up to Max tokens. A newly
+	// tracked key initially receives Max tokens.
+	Max int64
+	// Overdraft is the amount of additional tokens a key can consume
+	// once it's nominally run out of tokens, effectively going into
+	// debt. A key must repay its debt, by letting
+	// RefillEvery*Overdraft time elapse without requesting tokens,
+	// before tokens become available for consumption again.
+	//
+	// A non-zero overdraft results in "cooldown" behavior: with a
+	// normal token bucket that bottoms out at zero tokens, an abusive
+	// key can still consume one token every RefillEvery
+	// interval. With a non-zero overdraft, a throttled key must stop
+	// requesting tokens entirely for a cooldown period, otherwise
+	// they remain perpetually in debt and cannot proceed at all.
+	Overdraft int64
+	// RefillEvery is the interval at which a key regains tokens for
+	// use, up to Max tokens.
+	RefillEvery time.Duration
+
+	// even though expirable.LRU is thread-safe, it cannot atomically
+	// upsert, so we have to wrap it in our own mutex.
+	mu    sync.Mutex
+	cache *lru.Cache[K, *bucket]
```

Comment:
Is that BSL? We have an implementation in tailscale.com/util/lru that should work instead?

Reply:
It's still MPL, but TIL we have one. I'll switch to that.